valhalla/baldr/traffictile.h

```diff
@@ -0,0 +1,128 @@
+ *
+ * TileHeader (24 bytes)
+ * n x Speed entries (n x 4 bytes)
+ * IncidentsHeader (8 bytes)
+ * m x Incidents entries (buffer 0) (m x 8 bytes)
+ * m x Incidents entries (buffer 1) (m x 8 bytes)
+ *
+ * The Incidents are sparse - the TileHeader->incident_buffer_size specifies how
+ * much space is available in the tile.  The IncidentsHeader fields specify where
+ * incidents begin/end in this buffer.  This allows for new incident lists to be
+ * placed in the buffer, then the begin/end updated to atomicly move readers to
+ * a new location.
+ */
+class Tile {
+public:
+  Tile(const char* tile_ptr)
+      : header{reinterpret_cast<const TileHeader*>(tile_ptr)},
+        speeds{reinterpret_cast<const volatile Speed*>(tile_ptr + offsetof(Tile, speeds))},
+        incident_count_0{reinterpret_cast<const volatile std::uint32_t*>(
+            tile_ptr + offsetof(Tile, incident_count_0))},
+        incident_count_1{
+            reinterpret_cast<const volatile std::uint32_t*>(
+                tile_ptr + offsetof(Tile, incident_count_1))},
+        incidents_0{reinterpret_cast<const volatile Incident*>(
+            tile_ptr + offsetof(Tile, incidents_0))},
+        incidents_1{reinterpret_cast<const volatile Incident*>(
+            tile_ptr + offsetof(Tile, incidents_1))} {
+  }
+
+  Tile(const Tile& other) = default;
+  Tile& operator=(const Tile& other) = default;
+
+  const volatile Speed* getTrafficForDirectedEdge(const std::uint32_t directed_edge_offset) {
+    assert(directed_edge_offset < header->directededge_count);
+    return speeds + directed_edge_offset;
+  }
+
+  const std::vector<const Incident>
+  getIncidentsForDirectedEdge(const std::uint32_t directed_edge_offset) {
+    // Sanity check and exit early if false
+    if (!getTrafficForDirectedEdge(directed_edge_offset)->has_incident)
+      return {};
+
+    // Copy the current active buffer so it won't change during our usage below
+    auto active_buffer = header->active_incident_buffer;
+
+    const auto& count = active_buffer == 0 ? *incident_count_0 : *incident_count_1;
+    const auto& buffer = active_buffer == 0 ? incidents_0 : incidents_1;
+
+    // Find the beginning and end of edges that match our directed_edge_offset
+    // Note that equal_range uses a binary search when given a LegacyRandomAccessIterator,
+    // which raw pointers satisfy.
+    struct Comp {
+      bool operator()(const volatile Incident& incident, std::uint32_t edge_index) const {
+        return incident.edge_index < edge_index;
+      }
+      bool operator()(std::uint32_t edge_index, const volatile Incident& incident) const {
+        return edge_index < incident.edge_index;
+      }
+    };
+    const auto range = std::equal_range(buffer, buffer + count, directed_edge_offset, Comp{});
+
+    // Copy the results so that they're non-volatile for our caller
+    // TODO: race condition: make sure our range is still valid
+    return std::vector<const Incident>(range.first, range.second);
```

Comment:
Whats the use of `const` in the template argument?

Reply:
Removed, I was being over-zealous.  It didn't turn out to be practical.